drivers: add ethernet over serial driver

kaspar030 opened the pull request:
I needed a one-cable solution for connecting boards with both network and console. So I came up with this driver, which is essentially l2 SLIP and additional support for multiplexing STDIO over the same serial connection.

It is an alternative to gnrc_slip / tuntap.

Differences:
- it implements a netdev driver
- it transfers l2 ethernet frames instead of IP packets
- it enables network and stdio over just one serial cable
- it uses a slightly more robust framing
- it is more code
- it is only tested on Linux
- it uses more memory
- it has slower ping times

The last two will probably be fixed.

In order to play with it, see ```dist/tools/ethos/README.md``` for what to add to your makefiles.

OlegHahm commented:
Cool, that could help to run a 6LBR on the IoT-LAB testbed.

cladmi commented:
I also thought that the shell could also be available as a server. This way, the multiplexing would directly use the network stack. 

It would just needed to set the first ip address automatically and print in on the serial line like as done in contiki examples.

authmillenon commented:
did not look into your implementation in detail, but shouldn't it be way simpler to just multiplex slip and shell input/output over one UART?

kaspar030 commented:
> I also thought that the shell could also be available as a server. This way, the multiplexing would directly use the network stack.

... which in turn would make it impossible to debug the network stack.

kaspar030 commented:
> did not look into your implementation in detail, but shouldn't it be way simpler to just multiplex slip and shell input/output over one UART?

It would end up being the same but on Layer 3 in principal, which honestly I realized when taking a look at how you did some things in gnrc_slip.

But ethos is also netdev based and thus works without gnrc.

authmillenon commented:
Ah, I missed the non-gnrc part :-)

authmillenon commented on drivers/ethos/ethos.c:
```diff
@@ -0,0 +1,322 @@
+    tsrb_init(&dev->inbuf, (char*)buf, bufsize);
+    mutex_init(&dev->out_mutex);
+
+    uint32_t a = genrand_uint32();
```
I already hate the fact, that `netdev2_tap` has more or less random MAC addresses, but at least they are somewhat stable as long as you don't rebuild the TAP. Can't you use the CPUID to generate the MAC? It would make debugging and testing with these devices way easier ;-).

authmillenon replied:
Also keep in mind that bits 7 (U/L) and 8 (I/G) have special meaning in 48-bit MAC addresses http://standards.ieee.org/develop/regauth/tut/eui.pdf

authmillenon replied:
(address is globally unique if bit 7 is unset [so you need to set it, since you generated it], bit 8 needs to be unset since your address is a unicast address)

kaspar030 replied:
Totally. Will fix.

cgundogan replied:
@kaspar030 do you have time to address @authmillenon's comment until the Hack'n'Ack session?

emmanuelsearch commented:
Is this too WIP to be an H&A candidate? Arguably it is a new functionality that could be improved later too... right?

kYc0o commented:
Hey! I was about to ask the same, if I can ACK this PR let me know

cgundogan commented:
@emmanuelsearch I used this feature several times with great success. I would also vote for getting this merged as soon as @kaspar030 addressed @authmillenon's comments. I couldn't get `ethos` to work with the iotlab-testbed, though. I am not sure if this is due to `ethos` or whether it is some magic that I am unaware of when I tunnel the serial aggregator to my local machine in order to use it with `ethos`. @kaspar030 I'd like to talk to you about this f2f if you attend tomorrow's H&A.

emmanuelsearch commented:
@kYc0o did you already test this? On which hardware/config?

emmanuelsearch commented:
@cgundogan you used successfully this with SAMR21, I suppose?

cgundogan commented:
yes, the typical scenario: samr21-xpro running the border router example

kYc0o commented:
I didn't already test but I plan to do it as soon as possible on IoT-Lab

thomaseichinger commented on dist/tools/ethos/ethos.c:
```diff
@@ -0,0 +1,313 @@
+
+#include <netinet/in.h>
+#include <linux/if.h>
+#include <linux/if_tun.h>
```
Please mention in the README or somewhere else this is only usable with linux.

thomaseichinger commented on dist/tools/ethos/ethos.c:
```diff
@@ -0,0 +1,313 @@
+#include <errno.h>
```
License is missing.

thomaseichinger commented on dist/tools/ethos/ethos.c:
```diff
@@ -0,0 +1,313 @@
+    cfsetospeed (&tty, speed);
+    cfsetispeed (&tty, speed);
+
+    tty.c_cflag = (tty.c_cflag & ~CSIZE) | CS8;     // 8-bit chars
```
Here and all following, please change comments' style.

cgundogan commented:
interesting - tested with `iotlab-m3` and I didn't receive any output from `ethos`. Tried with `samr21-xpro` afterwards and there was also no output from `ethos`. Will check more thoroughly tomorrow

kaspar030 commented:
- addressed comments

kaspar030 commented:
I had some trouble using this on a uclibc-based OpenWrt as host.

kaspar030 commented:
I just realized that ethos doesn't work until I've used the board once using "make term". Probably something is missing in ethos' linux side terminal setup.

authmillenon commented:
The tap interface isn't usually "active" until some software is running on it. Maybe that's it?

emmanuelsearch commented:
so is this actually being worked on as H&A candidate or not?

authmillenon commented:
@kaspar030 just came in :-)

kaspar030 commented:
Well, I initialized the serial port using "115200" as speed value, posix expects a define ("B115200"), so it was pure luck that ethos ever worked at all. :)
Fixed...

kaspar030 commented:
I'm using this quite a lot, so I don't consider it WIP anymore.

cgundogan commented:
>I'm using this quite a lot, so I don't consider it WIP anymore.

+1. BTW, @kaspar030 can you verify whether ethos works for uclibc with your fix from yesterday?

emmanuelsearch commented:
OK, so if there are already multiple happy users, then an ACK should appear soonish I suppose and merging should follow quickly?

emmanuelsearch commented:
mishap close, sorry (so immediate reopen)

kYc0o commented:
@kaspar030 can you give more details about your tests? I'm having difficulties to make this work... the interface seems to work using ethos but when I try to initialize a network by adding a prefix to the radio interface it freezes if another board is running, for instance a samr21 running gnrc_networking... we thought it was related to #4705 but now I don't know what is happening

kaspar030 commented:
@kYc0o I'm using two samr21, one running gnrc_border_router with ethos, the other running vanilla gnrc_networking. I don't get freezes, just experiencing routing problems, but I'm also hacking on RIOT's RA handling.

kaspar030 commented:
> @kaspar030 can you verify whether ethos works for uclibc with your fix from yesterday?

Confirmed, I'm borderrouting from OpenWrt... ;)

cgundogan commented:
>Confirmed, I'm borderrouting from OpenWrt... ;)
eureka!

kaspar030 commented:
- linux ethos now exits when reading from serial produces an error (instead of eating cpu)
- fixed a bug on the linux side: if RIOT would send a frame delimiter and then crash, the in/out frame state on the linux side was reversed. fixed.

OlegHahm commented:
I have some requests for improvement:
* some startup message would be helpful (since one has to run sudo and it's difficult if the program started or one mistyped the password)
* adding readline would also make things easier
* add to the README that the shell runs inside the ethos tool

In a next step we may consider adapting the 6LBR example to ethos.

OlegHahm commented:
@cgundogan, a configuration that works:
On 6LBR:
```
ifconfig 6 add affe::2/48
ifconfig 6 add affe::2/48
success: added affe::2/48 to interface 6
> ifconfig 5 add affe::3/64
ifconfig 5 add affe::3/64
success: added affe::3/64 to interface 5
> fibroute add :: via <link-local of tap> dev 6
fibroute add :: via <link-local of tap> dev 6
```
on Linux
```
> sudo ip link set tap0 up
> sudo ip a a affe::1/48 dev tap0
> sudo ip r d affe::/48 dev tap0
> sudo ip r a affe::2 dev tap0
sudo ip r a affe::/48 via affe::2 dev tap0
```

OlegHahm commented:
After interrupting the RIOT application with SIGINT from gdb, firing a command on the RIOT shell and continuing the app, the ethos tool fails to communicate - even after restarting it.

cgundogan commented:
Neat! Will test that tomorrow. Thanks.
Am 29.01.2016 20:14 schrieb "Oleg Hahm" <notifications@github.com>:

> @cgundogan <https://github.com/cgundogan>, a configuration that works:
> On 6LBR:
>
> ifconfig 6 add affe::2/48
> ifconfig 6 add affe::2/48
> success: added affe::2/48 to interface 6
> > ifconfig 5 add affe::3/64
> ifconfig 5 add affe::3/64
> success: added affe::3/64 to interface 5
> > fibroute add :: via <link-local of tap> dev 6
> fibroute add :: via <link-local of tap> dev 6
>
> on Linux
>
> > sudo ip link set tap0 up
> > sudo ip a a affe::1/48 dev tap0
> > sudo ip r d affe::/48 dev tap0
> > sudo ip r a affe::2 dev tap0
> sudo ip r a affe::/48 via affe::2 dev tap0
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/4438#issuecomment-176918548>.
>

cgundogan commented:
Did you merge in the latest commit of @kaspar030 which fixes some bugs that
sound similar to your stated ones?
Am 29.01.2016 20:19 schrieb "Oleg Hahm" <notifications@github.com>:

> After interrupting the RIOT application with SIGINT from gdb, firing a
> command on the RIOT shell and continuing the app, the ethos tool fails to
> communicate - even after restarting it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/4438#issuecomment-176919938>.
>

OlegHahm commented:
I did.

kaspar030 commented:
> After interrupting the RIOT application with SIGINT from gdb, firing a command on the RIOT shell and continuing the app, the ethos tool fails to communicate - even after restarting it.

I can't reproduce this. Do you have a sequence of actions that does? This might be another state problem within ethos.

kaspar030 commented:
I noticed that a simple reboot of RIOT would mess up Linux's neighbor cache information as RIOT's L2 address changes. So now ethos sends (and replies to) HELLO frames, and the Linux ethos tool clears it's tap interface's neighbor cache.

OlegHahm commented:
I'll try to provide such a sequence. Remind me on Wednesday if I forget.

OlegHahm commented:
Cannot reproduce the error any more. (Thanks, @kYc0o, for the reminder.) ACK for the functionality of this PR. (readline would be still nice to have. ;))

kYc0o commented:
I ACK this PR too. However, please notice this https://lists.riot-os.org/pipermail/devel/2016-February/003386.html since I had the same issue.

kaspar030 commented:
> I ACK this PR too. However, please notice this https://lists.riot-os.org/pipermail/devel/2016-February/003386.html since I had the same issue.

RIOT doesn't answer to pings for addresses that were not the interface the echo request came in. See #4456.

OlegHahm commented:
Confirmed.

kaspar030 commented:
- fixed stupid bug on linux side, rebased

kYc0o commented:
Now is also working on my side. I think if travis is happy the PR can be merged

kaspar030 commented:
I'll have to squash first...

kaspar030 commented:
- squashed

kaspar030 commented:
Travis seem lazy. :)

OlegHahm commented:
`>file has an unknown license header: 'dist/tools/ethos/ethos.c'`

kaspar030 commented:
- fixed license header problem, doxygen, static code checker nitpicks

kaspar030 commented:
Mr. T. is all green. Anyone cares to say ACK and press merge?

kYc0o commented:
ACK :)
